### DISCUSSION TITLE
Pagination on raw table and custom sql query with MS SQL Server

### DISCUSSION THREAD

**Discussion by EvanGllt** [Budibase/budibase#14208] [Help] [answered]
Hello, there!

How do I apply pagination to my raw boards? Because no matter how hard I look at the official documentation, it doesn't apply.

![Aide 1](https://github.com/user-attachments/assets/460fc42f-f52b-4014-bbe1-c85cd3c8c314)
![Aide 2](https://github.com/user-attachments/assets/426def67-c2ef-4244-b2d1-a8b66cee032d)

Knowing that I have the same problem for custom requests. I know it's not possible to do this directly, but I can't find anything to help me. I found this (https://github.com/Budibase/budibase/discussions/7313#discussioncomment-3430353) but it didn't work...

I use a MS SQL Server data source.

Can anyone help me? Thanks in advance.

**Comment by ConorWebb96** [comment]
Hey @EvanGllt,

You may want to look into `limit` and `offset` for your [SQL queries](https://www.sqltutorial.org/sql-limit/) (I've attached some additional reading you can do on this, that may aid you). You could then attach bindings to these values and then cycle through the pages with buttons that amend the bindings related to your query.

Hopefully this makes sense

**Reply by EvanGllt** [reply]
Hello and thank you for your reply.

I've already tried LIMIT and OFFSET and it doesn't work. I've seen that it's not possible with MS SQL SERVER but that you have to use: OFFSET {{min}} ROWS FETCH NEXT {{max}} ROWS ONLY but even that doesn't work.
I get stuck at the custom query validation stage :

![Aide 1](https://github.com/user-attachments/assets/8ee5eadc-2aa5-43a2-a4ed-4580ba285f17)

I'm crying.

**Reply by ConorWebb96** [reply]
I can't see you're entire query. But heres a more simple version of a query I wrote a while ago that works for pagination.

```sql
SELECT *
FROM persons
WHERE gender = 'female'
LIMIT {{limits}} OFFSET {{offset}};
```

Demo of this working.
![20240723_142509](https://github.com/user-attachments/assets/4045b9a8-89f6-4f78-8b21-5189a93b06ae)

The only main difference is this is a MySQL database compared to your MSSQL database. Does your current query work outside of Budibase, stripping out all of the ``{{ }}`` syntax?

**Reply by EvanGllt** [reply]
Removing the {{ }} braces works, but with the bindings I get an error asking for an int value. Except I can see how I can get around this if I need the bindings. That's weird.

![Aide 1](https://github.com/user-attachments/assets/74a3858c-2cf6-441c-8e33-8485a68d814d)
![Aide 2](https://github.com/user-attachments/assets/8ab92634-71f9-479e-8a32-32641cd1a259)

**Reply by ConorWebb96** [reply]
Since its related to a string being passed into it. Why don't you click the lighting bolt icon switch to JS and return 0 or 5 that way to ensure its a number being passed?

```js
return 0
```

I think {{ literal 0 }} in hbs should work as well?

**Reply by EvanGllt** [reply]
I had no choice but to use the text type for bindings :

![Aide 2](https://github.com/user-attachments/assets/064e37b7-b5fb-49a1-bc21-972322fe6fce)

And `return 0` doesn't change anything :

![Aide 1](https://github.com/user-attachments/assets/14c1e5dd-3b93-422a-8576-55b57afcb565)

**Reply by ConorWebb96** [reply]
Ah, my mistake, I thought JS was available. You may need to use CAST or CONVERT to ensure that the values being provided are integers.

**Reply by EvanGllt** [reply]
It works, but I don't know why the first four rows are duplicated when I only have one row like 

![Aide 3](https://github.com/user-attachments/assets/aa3f38f9-1ab8-44cc-a202-34e02331c7e0)

![Aide 1](https://github.com/user-attachments/assets/946180ee-10c3-422e-bf3d-96a0de31feda)

![Aide 2](https://github.com/user-attachments/assets/e6053968-d235-4823-8388-bd7461c2ae27)

**Comment by EvanGllt** [comment, marked as answer]
Hi,

I found out how to do it.

Here's how I did it for people who are interested :

![Aide 0](https://github.com/user-attachments/assets/46b8393e-b6a9-47c0-b697-dc4d3e918f7c)

For my SQL query I added a `DISTINCT` after `SELECT` :

![Aide 1](https://github.com/user-attachments/assets/149789c8-eacb-4c74-b5bf-64c50ab2103a)

I have this component tree :

![Aide 2](https://github.com/user-attachments/assets/cd88d31b-adf8-42be-bd1e-878037a608f2)

That's bindings there :

![Aide 3](https://github.com/user-attachments/assets/156fb9cd-54a4-40a7-a8c8-c9c24bed2281)

This line in javascript for `offset` :

![Aide 4](https://github.com/user-attachments/assets/4e3fcd02-c5c4-4dc4-801f-4f1238766c56)

For the `Back Button` :

![Aide 5](https://github.com/user-attachments/assets/51a0ff99-7607-4d1f-a354-7c25927fb36b)
![Aide 6](https://github.com/user-attachments/assets/b1d4d441-d604-4d4f-9ec1-ec599f8cb385)

For the `Next Button` :

![Aide 7](https://github.com/user-attachments/assets/4de1d2c5-18e6-49cd-a1ca-a139a1724fa7)
![Aide 8](https://github.com/user-attachments/assets/1afe213a-93ac-4a5c-af5d-95139018d482)

And for `Page Field` :

![Aide 9](https://github.com/user-attachments/assets/c6b504ae-1dad-4888-84cc-5e6bcbac5019)
